// write-excel-file/commonjs/xlsx/files/sheet.xml/sheet.xml.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = generateSheetXml;
var _rows = _interopRequireDefault(require("./rows.js"));
var _columns = _interopRequireDefault(require("./columns.js"));
var _processMergedCells2 = _interopRequireDefault(require("./processMergedCells.js"));
var _mergedCellsDescription = _interopRequireDefault(require("./mergedCellsDescription.js"));
var _layout = _interopRequireDefault(require("./layout.js"));
var _views = _interopRequireDefault(require("./views.js"));
var _drawingReference = _interopRequireDefault(require("./drawingReference.js"));
var _getAdditionalContent = _interopRequireDefault(require("../../helpers/features/getAdditionalContent.js"));
var _transformContent = _interopRequireDefault(require("../../helpers/features/transformContent.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function generateSheetXml(sheetXmlParameters, features) {
  var sheetData_ = sheetXmlParameters.sheetData,
    sheetOptions = sheetXmlParameters.sheetOptions,
    sheetIndex = sheetXmlParameters.sheetIndex,
    sheetId = sheetXmlParameters.sheetId,
    hasDefaultFont = sheetXmlParameters.hasDefaultFont,
    findOrCreateCellStyle = sheetXmlParameters.findOrCreateCellStyle,
    findOrCreateSharedString = sheetXmlParameters.findOrCreateSharedString;
  var columns = sheetOptions.columns,
    dateFormat = sheetOptions.dateFormat,
    orientation = sheetOptions.orientation,
    showGridLines = sheetOptions.showGridLines,
    rightToLeft = sheetOptions.rightToLeft,
    zoomScale = sheetOptions.zoomScale;
  var _processMergedCells = (0, _processMergedCells2["default"])(sheetData_, {
      features: features
    }),
    sheetData = _processMergedCells.sheetData,
    mergedCells = _processMergedCells.mergedCells;
  var xml = '<?xml version="1.0" ?>' + '<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main">' +
  // For some weird reason, Excel 2007 demands `<sheetViews/>` element to be the first one.
  (0, _views["default"])({
    showGridLines: showGridLines,
    rightToLeft: rightToLeft,
    zoomScale: zoomScale,
    sheetIndex: sheetIndex
  }) +
  // For some weird reason, Excel 2007 demands `<cols/>` element to follow `<sheetViews/>` element.
  (0, _columns["default"])(columns) + '<sheetData>' + (0, _rows["default"])(sheetData, {
    findOrCreateCellStyle: findOrCreateCellStyle,
    findOrCreateSharedString: findOrCreateSharedString,
    hasDefaultFont: hasDefaultFont,
    features: features,
    dateFormat: dateFormat
  }) + '</sheetData>' + (0, _mergedCellsDescription["default"])(mergedCells) + (0, _layout["default"])({
    sheetId: sheetId,
    orientation: orientation
  }) + (0, _drawingReference["default"])() +
  // Apply any plugins that insert additional content to this XML.
  (0, _getAdditionalContent["default"])('xl/worksheets/sheet{id}.xml', features, sheetOptions, {
    sheetIndex: sheetIndex,
    sheetId: sheetId
  }) + '</worksheet>';

  // Apply any plugins that transform this XML.
  xml = (0, _transformContent["default"])(xml, 'xl/worksheets/sheet{id}.xml', features, sheetOptions, {
    sheetIndex: sheetIndex,
    sheetId: sheetId
  });
  return xml;
}
